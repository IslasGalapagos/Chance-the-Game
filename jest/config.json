{
  "rootDir": "../",
  "moduleFileExtensions": ["ts", "tsx", "js", "jsx", "json", "node"],
  "setupFiles": ["<rootDir>/jest/setup.js"],
  "testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
  "transform": {
    "^.+\\.tsx?$": "ts-jest"
  }
}
