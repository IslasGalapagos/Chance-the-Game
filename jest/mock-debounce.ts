/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

type TDebouncedF = () => {};

let debouncedF: TDebouncedF;

export const debouncedMock = jest.fn(() => debouncedF());

jest.mock('lodash-es/debounce', () =>
  jest.fn((fn: TDebouncedF) => {
    debouncedF = fn;
    return debouncedMock;
  })
);
