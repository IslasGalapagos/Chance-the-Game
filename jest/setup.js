import {configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import jsdom from 'jsdom';

configure({adapter: new Adapter()});

const {JSDOM} = jsdom;
const dom = new JSDOM(
  '<!DOCTYPE html><html><body><div id="root"></div></body></html>'
);

window.eval(`
  Object.defineProperty(window.document.body, 'clientHeight', {
    get() {
      return window.innerHeight;
    }
  });
`);

global.window = window;
global.document = window.document;

global.window.resizeTo = (width, height) => {
  global.window.innerWidth = width || global.window.innerWidth;
  global.window.innerHeight = width || global.window.innerHeight;
  global.window.dispatchEvent(new Event('resize'));
};
