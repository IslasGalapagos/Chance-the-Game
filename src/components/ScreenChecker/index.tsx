/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import debounce from 'lodash-es/debounce';
import StyledStub from './styles';

interface MinSizes {
  readonly WIDTH: number;
  readonly HEIGHT: number;
}

export const minSizes: Readonly<MinSizes> = {
  WIDTH: 960,
  HEIGHT: 740
};

interface State {
  readonly isSuitable: boolean;
}

class ScreenChecker extends React.PureComponent<{}, State> {
  readonly state: State;

  onResize = debounce(() => {
    this.setState({
      isSuitable: this.check()
    });
  }, 200);

  constructor(props: any) {
    super(props);

    this.state = {
      isSuitable: this.check()
    };
  }

  check(): boolean {
    const windowWidth: number = window.innerWidth;
    const windowHeight: number = document.body.clientHeight;
    const {WIDTH, HEIGHT} = minSizes;

    if (windowWidth < WIDTH || windowHeight < HEIGHT) {
      return false;
    }

    return true;
  }

  componentDidMount() {
    window.addEventListener('resize', this.onResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  render() {
    const {isSuitable} = this.state;
    return (
      <div>
        {!isSuitable ? (
          <StyledStub>
            <span>
              Sorry, but your screen sizes are smaller than required.
              <br />
              Minimum needed: width=960, height=800
            </span>
          </StyledStub>
        ) : null}
      </div>
    );
  }
}

export default ScreenChecker;
