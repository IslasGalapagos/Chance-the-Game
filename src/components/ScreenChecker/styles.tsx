/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import styled from 'react-emotion';

const StyledStub = styled('div')`
  position: absolute;
  left: 0;
  top: 0;
  z-index: 5;

  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;
  height: 100%;

  background: #0000b1;

  span {
    font-family: 'Courier New', Courer;
    font-size: 25px;
    font-weight: bold;
    text-align: center;

    padding: 0 30px;

    color: #fff;

    @media (min-width: 740px) {
      padding: 0;
    }
  }
`;

export default StyledStub;
