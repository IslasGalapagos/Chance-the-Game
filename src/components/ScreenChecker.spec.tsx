/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import {shallow} from 'enzyme';
import * as React from 'react';
import * as testRenderer from 'react-test-renderer';
import debounce from 'lodash-es/debounce';
import {debouncedMock} from '../../jest/mock-debounce';
import ScreenChecker, {minSizes} from './ScreenChecker';
import Stub from './ScreenChecker.styles';

describe('<ScreenChecker/>', () => {
  beforeEach(() => {
    (debounce as any).mockClear();
    debouncedMock.mockClear();
  });

  it('renders correctly', () => {
    window.resizeTo(minSizes.WIDTH - 1, minSizes.HEIGHT);
    const treeAndStyles = testRenderer.create(<ScreenChecker />).toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  // THIS.MINSIZES
  it('has minSizes', () => {
    expect(minSizes).toEqual(expect.objectContaining({
      WIDTH: expect.any(Number),
      HEIGHT: expect.any(Number)
    }));
  });

  // STATE
  it(`sets bool dependent on ScreenChecker.check's result to state.isSuitable`, () => {
    window.resizeTo(minSizes.WIDTH, minSizes.HEIGHT);

    const spyCheck = jest.spyOn(ScreenChecker.prototype, 'check');
    const component = shallow(<ScreenChecker />);
    expect(spyCheck).toBeCalled();
    expect(spyCheck.mock.results[0].value).toEqual(true);
    expect(component.state('isSuitable')).toEqual(true);

    spyCheck.mockRestore();
  });

  // METHODS
  it('has check that returns bool dependent on window sizes', () => {
    const component = shallow(<ScreenChecker />);
    const {check} = component.instance() as ScreenChecker;
    window.resizeTo(minSizes.WIDTH, minSizes.HEIGHT);
    expect(check()).toEqual(true);
    window.resizeTo(minSizes.WIDTH - 1, minSizes.HEIGHT);
    expect(check()).toEqual(false);
    window.resizeTo(minSizes.WIDTH, minSizes.HEIGHT - 1);
    expect(check()).toEqual(false);
  });

  it('has onResize that calls ScreenChecker.check and sets result to state.isSuitable', () => {
    window.resizeTo(minSizes.WIDTH, minSizes.HEIGHT);

    const spyCheck = jest.spyOn(ScreenChecker.prototype, 'check');
    const component = shallow(<ScreenChecker />);
    expect(component.state('isSuitable')).toEqual(true);
    spyCheck.mockClear();
    const {onResize} = component.instance() as ScreenChecker;
    window.resizeTo(minSizes.WIDTH - 1, minSizes.HEIGHT);
    onResize();
    expect(spyCheck).toBeCalled();
    expect(component.state('isSuitable')).toEqual(false);

    spyCheck.mockRestore();
  });

  it('calls ScreenChecker.check in onResize with debounce', () => {
    const spyCheck = jest.spyOn(ScreenChecker.prototype, 'check');
    expect(debounce).not.toHaveBeenCalled();
    const {onResize} = shallow(<ScreenChecker />).instance() as ScreenChecker;
    expect(debounce).toHaveBeenCalledTimes(1);
    expect(debouncedMock).not.toHaveBeenCalled();
    spyCheck.mockClear();
    onResize();
    expect(debouncedMock).toHaveBeenCalledTimes(1);
    expect(spyCheck).toHaveBeenCalledTimes(1);

    spyCheck.mockRestore();
  });

  // EVENT HANDLERS
  it('sets ScreenChecker.onResize to window.onresize handler on mount', () => {
    const spyCDM = jest.spyOn(ScreenChecker.prototype, 'componentDidMount');
    const spyWindowAEL = jest.spyOn(window, 'addEventListener');
    const {onResize} = shallow(<ScreenChecker />).instance() as ScreenChecker;
    expect(spyCDM).toBeCalledTimes(1);
    expect(spyWindowAEL).toBeCalledTimes(1);
    expect(spyWindowAEL.mock.calls[0][0]).toEqual('resize');
    expect(spyWindowAEL.mock.calls[0][1]).toEqual(onResize);

    spyCDM.mockRestore();
    spyWindowAEL.mockRestore();
  });

  it('remove ScreenChecker.onResize from window.onresize handler on unmount', () => {
    const spyCDU = jest.spyOn(ScreenChecker.prototype, 'componentWillUnmount');
    const spyWindowREL = jest.spyOn(window, 'removeEventListener');
    const component = shallow(<ScreenChecker />);
    const {onResize} = component.instance() as ScreenChecker;
    component.unmount();
    expect(spyCDU).toBeCalledTimes(1);
    expect(spyWindowREL).toBeCalledTimes(1);
    expect(spyWindowREL.mock.calls[0][0]).toEqual('resize');
    expect(spyWindowREL.mock.calls[0][1]).toEqual(onResize);

    spyCDU.mockRestore();
    spyWindowREL.mockRestore();
  });

  // STUB
  it('renders Stub when state.isSuitable is false, otherwise returns null', () => {
    const component = shallow(<ScreenChecker />);
    component.setState({isSuitable: false});
    expect(component.find(Stub).length).toEqual(1);
    component.setState({isSuitable: true});
    expect(component.find(Stub).length).toEqual(0);
  });
});
