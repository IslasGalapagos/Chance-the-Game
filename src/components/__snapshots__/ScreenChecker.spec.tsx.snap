// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScreenChecker/> renders correctly 1`] = `
.emotion-0 {
  position: absolute;
  left: 0;
  top: 0;
  z-index: 4;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background: #0000b1;
}

.emotion-0 span {
  font-family: 'Courier New',Courer;
  font-size: 25px;
  font-weight: bold;
  text-align: center;
  padding: 0 30px;
  color: #fff;
}

@media (min-width:740px) {
  .emotion-0 span {
    padding: 0;
  }
}

<div>
  <div
    className="emotion-0"
  >
    <span>
      Sorry, but your screen size is smaller than required.
      <br />
      Minimal sizes: width=960, height=800
    </span>
  </div>
</div>
`;
